### ISSUE TITLE
Basic example with express: Typescript fails with `Type instantiation is excessively deep and possibly infinite`

### ISSUE THREAD

**Issue by ships** [ts-rest/ts-rest#389]
Hi friends, I am exploring ts-rest and hope not to use Zod directly, as I use `io-ts` for a lot of similar typing work. I am following the quickstart and have installed Zod as a dev dependency. Here are my relevant versions:
```
  "dependencies": {
    "@ts-rest/core": "^3.30.2",
    "@ts-rest/express": "^3.30.2",
    "express": "^4.18.2"
  },
  "devDependencies": {
    "@types/express": "^4.17.17",
    "typescript": "^4.9.5",
    "zod": "^3.22.2"
  },
```

However, when I contrive this basic example contract and server:
```
// contract.ts
const c = initContract()

export const contract = c.router({
  getInfo: {
    method: 'GET',
    path: '/info',
    responses: {
      200: c.type<ApiInfo>(),
    },
    summary: 'Get information about this Docmaps API server',
  },
})

// server.ts
  const s = initServer()

  const router = s.router(contract, {
    getInfo: async () => {
      const info = api.get_info()

      return {
        status: 200,
        body: info,
      }
    },
  })

  createExpressEndpoints(contract, router, app)        // <---- ERROR
```

I get this error on typescript compilation: `Type instantiation is excessively deep and possibly infinite. typescript (2589)`

Apparently this is an issue that occurs in certain uses of Zod, but since I don't invoke Zod, i expect this might be an issue with how ts-rest calls Zod.

Can you help me understand if there is a known workaround for it while using ts-rest?

**Comment by ships** [comment]
I also tried this with removing `zod` as an additional dependency, and with it in prod dependencies rather than dev; but the error persists.

**Comment by Talent30** [comment]
Please upgrade to zod 3.1.4

**Comment by ships** [comment]
Thanks for the suggestion! After this downgrade I am still seeing the issue.

**Comment by Talent30** [comment]
> Thanks for the suggestion! After this downgrade I am still seeing the issue.

We had the same issue, with zod 3.22.x, could you try 

a) downgrade your **ALL** zod if you are using monorepo and then
b) delete your node_modules folder and
c) npm i
d) if above didn not work please delete package-lock.json as well and then run npm i

This is how we solved it.

Something might be related https://github.com/colinhacks/zod/issues/2697

**Comment by Irelynx** [comment]
Hi. I have the same problem after updating ts-rest library.
Downgrading zod doesn't help in my situation, but downgrading @ts-rest/core to 3.28.0 fixes the issue (with zod@3.22.2)

**Comment by ships** [comment]
Thanks @Irelynx and @Talent30 . I also can confirm that downgrading Zod did not help, but downgrading @ts-rest/core to 3.28.0 did prevent the error.

**Comment by Talent30** [comment]
Hey @ships. Looks like it is working with typescript 5.3 beta

**Comment by michaelangeloio** [comment]
@Talent30 @ships @Irelynx is this still an issue? does anyone have a minimal reproduction repo?

**Comment by Talent30** [comment]
> @Talent30 @ships @Irelynx is this still an issue? does anyone have a minimal reproduction repo?

Hi, @michaelangeloio. I can’t reproduce it now.

**Comment by ships** [comment]
I haven't had a chance to upgrade Typescript but will report back if that solves the issue, it should come in a week or two on this project.

**Comment by miku86** [comment]
> @Talent30 @ships @Irelynx is this still an issue? does anyone have a minimal reproduction repo?

I was able to solve it by:
1. updating typescript from `5.2.2` to `5.3.0-beta`
2. removing `package-lock.json`
3. removing `node_modules`
4. running `npm install`

**Comment by ships** [comment]
Okay, i finally got round to this typescript-5 upgrade and can confirm that as of ts-5.2.2 I no longer have this issue. I didn't even need to upgrade to typescript 5.3 beta.

**Comment by muellerluke** [comment]
Anyone else experiencing this still? I'm on express 4.21.2 @ts-rest/express 3.52.1, and typescript 5.8.3 on the backend and @ts-rest/core 3.5.21 and typescript 5.8.3 on the frontend.

**Comment by UBy** [comment]
I just ran into this issue.
TS: 5.8.3
ts-rest: 3.52.1
zod: 3.25.62

**Comment by MZPL** [comment]
I had a similar issue, but on the client side. I'm using a monorepo and share the contract in a separate package. During testin, `initClient()` worked fine when called inside the shared package, but errored on the client.

```
error TS2589: Type instantiation is excessively deep and possibly infinite
```
```
Argument of type '{ getConfig: { method: "GET"; path: "/getconfig"; responses: { 200: ZodObject<{ userChannelBlockOtherMessages: ZodBoolean; userChannelId: ZodNullable<ZodString>; canSendMessages: ZodBoolean; }, "strip", ZodTypeAny, { ...; }, { ...; }>; }; strictStatusCodes: true; }; sendChatMessage: { ...; }; sendDiscordMessage: { ....' is not assignable to parameter of type 'AppRouter'.
  Property 'getConfig' is incompatible with index signature.
    Type '{ method: "GET"; path: "/getconfig"; responses: { 200: ZodObject<{ userChannelBlockOtherMessages: ZodBoolean; userChannelId: ZodNullable<ZodString>; canSendMessages: ZodBoolean; }, "strip", ZodTypeAny, { ...; }, { ...; }>; }; strictStatusCodes: true; }' is not assignable to type 'AppRouter | AppRoute'.
      Type '{ method: "GET"; path: "/getconfig"; responses: { 200: ZodObject<{ userChannelBlockOtherMessages: ZodBoolean; userChannelId: ZodNullable<ZodString>; canSendMessages: ZodBoolean; }, "strip", ZodTypeAny, { ...; }, { ...; }>; }; strictStatusCodes: true; }' is not assignable to type 'AppRouteQuery'.
        Type '{ method: "GET"; path: "/getconfig"; responses: { 200: ZodObject<{ userChannelBlockOtherMessages: ZodBoolean; userChannelId: ZodNullable<ZodString>; canSendMessages: ZodBoolean; }, "strip", ZodTypeAny, { ...; }, { ...; }>; }; strictStatusCodes: true; }' is not assignable to type 'AppRouteCommon'.
          Types of property 'responses' are incompatible.
            Type '{ 200: ZodObject<{ userChannelBlockOtherMessages: ZodBoolean; userChannelId: ZodNullable<ZodString>; canSendMessages: ZodBoolean; }, "strip", ZodTypeAny, { ...; }, { ...; }>; }' is not assignable to type 'Record<number, AppRouteResponse>'.
              Property '200' is incompatible with index signature.
                Type 'ZodObject<{ userChannelBlockOtherMessages: ZodBoolean; userChannelId: ZodNullable<ZodString>; canSendMessages: ZodBoolean; }, "strip", ZodTypeAny, { ...; }, { ...; }>' is not assignable to type 'AppRouteResponse'.
                  Type 'ZodObject<{ userChannelBlockOtherMessages: ZodBoolean; userChannelId: ZodNullable<ZodString>; canSendMessages: ZodBoolean; }, "strip", ZodTypeAny, { ...; }, { ...; }>' is not assignable to type 'ZodType<any, ZodTypeDef, any>'.
                    The types returned by '_refinement(...)' are incompatible between these types.
                      Type 'ZodEffects<ZodObject<{ userChannelBlockOtherMessages: ZodBoolean; userChannelId: ZodNullable<ZodString>; canSendMessages: ZodBoolean; }, "strip", ZodTypeAny, { ...; }, { ...; }>, { ...; }, { ...; }>' is missing the following properties from type 'ZodEffects<ZodType<any, ZodTypeDef, any>, any, any>': _type, _output, _input, _def, and 33 more.
```

I solved it by downgrading zod to 3.25.10, which seems to be the latest working version. This happened completely at random for me and took some time to debug.
